Clean user agent to reduce metrics cardinality

dominikschulz opened the pull request:
**What this PR does / why we need it**:

This PR is an example implementation for my issue #31781.

``` release-note
```

This commit cleans common browser user-agents to reduce the metrics
cardinality in exported prometheus metrics.

Resolves kubernetes/kubernetes#31781

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31783)

<!-- Reviewable:end -->

lavalamp commented:
ok to test

lavalamp commented:
Can you sign the CLA? this LGTM as is.

@deads2k do you guys do anything with user agent that this will mess up?

deads2k commented:
> @deads2k do you guys do anything with user agent that this will mess up?

Thanks for the call out.  We do make use of the user agent in a couple spots, but this mutation is contained such that it won't cause us problems.

I am a little surprised to see it stripped so aggressively though.  Pretty much everyone includes that these days, right?  In addition to whatever browser it actually was later in the string.  Seems like this would squash firefox, chrome, and I think safari.  If I found the right doc, this might even squash down IE, right? https://msdn.microsoft.com/en-us/library/ms537503(v=vs.85).aspx

lavalamp commented:
If it's that aggressive, replacing it with `browser` might be better. Not
sure if we need to be more clear than that?

On Thu, Sep 1, 2016 at 6:09 AM, David Eads <notifications@github.com> wrote:

> @deads2k <https://github.com/deads2k> do you guys do anything with user
> agent that this will mess up?
>
> Thanks for the call out. We do make use of the user agent in a couple
> spots, but this mutation is contained such that it won't cause us problems.
>
> I am a little surprised to see it stripped so aggressively though. Pretty
> much everyone includes that these days, right? In addition to whatever
> browser it actually was later in the string. Seems like this would squash
> firefox, chrome, and I think safari. If I found the right doc, this might
> even squash down IE, right? https://msdn.microsoft.com/en-
> us/library/ms537503(v=vs.85).aspx
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31783#issuecomment-244073183>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAngll6XKx3UfS_vzPtcMz3WSIq0NQNuks5qls5vgaJpZM4JxlQj>
> .
>

deads2k commented:
> If it's that aggressive, replacing it with `browser` might be better. Not
> sure if we need to be more clear than that?

Replacing it with `Browser` sounds like a good idea.

dominikschulz commented:
Changed replacement to "Browser" as per @deads2k suggestion and signed CLA.

Thanks for considering this PR.

deads2k commented:
> Changed replacement to "Browser" as per @deads2k suggestion and signed CLA.
> 
> Thanks for considering this PR.

lgtm.  Went ahead and tagged it since lavalamp already said it was ok by him too.

deads2k commented:
@k8s-bot test this issue: #IGNORE

wojtek-t commented:
@k8s-bot test this, issue: #IGNORE ( Error in queue.)

dominikschulz commented:
Any change to get this merged? It's a rather trivial PR.

deads2k commented:
@k8s-bot test this, issue: #IGNORE ( Error in queue.)

deads2k commented:
> Any change to get this merged? It's a rather trivial PR.

real failures in verify

rmmh commented on pkg/apiserver/metrics/metrics_test.go:
```diff
@@ -0,0 +1,23 @@
+	}{
+		{
+			In:  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36",
+			Out: "Mozilla",
```
This should be `Out: "Browser"`

rmmh commented:
@dominikschulz this should be able to go in once the test case is fixed

mml commented on pkg/apiserver/metrics/metrics.go:
```diff
@@ -99,10 +100,17 @@ func InstrumentRouteFunc(verb, resource string, routeFunc restful.RouteFunction)
 
 		routeFunc(request, response)
-		Monitor(&verb, &resource, utilnet.GetHTTPClient(request.Request), rw.Header().Get("Content-Type"), delegate.status, now)
+		Monitor(&verb, &resource, cleanUserAgent(utilnet.GetHTTPClient(request.Request)), rw.Header().Get("Content-Type"), delegate.status, now)
```
Two suggestions.

1. Could we control this behavior with a flag?  I can imagine admins might want to turn it off to debug something.
2. Can we add an extra label to the metric explaining whether or not the client field has been canonicalized?  I vote for a string `clientType` which is either `raw` or `canonicalized`, but we could also do this with a bool.

dominikschulz replied:
You could make this depend on a flag, but if you try to turn these metrics into an debugging mechanism you're probably using them wrong.

The same goes for an extra label to tell it has been canonicalized.

If this is necessary to get this merged, ok, but I really think we shouldn't expose our users Prometheus servers to a huger number of (unnecessary) timeseries.

smarterclayton commented:
Could we just output a different field instead of client for canonical
clients?

On Jan 26, 2017, at 6:19 PM, Matt Liggett <notifications@github.com> wrote:

*@mml* requested changes on this pull request.
------------------------------

In pkg/apiserver/metrics/metrics.go
<https://github.com/kubernetes/kubernetes/pull/31783#pullrequestreview-18748752>
:

> @@ -99,10 +100,17 @@ func InstrumentRouteFunc(verb, resource string, routeFunc restful.RouteFunction)
 		response.ResponseWriter = rw

 		routeFunc(request, response)
-		Monitor(&verb, &resource, utilnet.GetHTTPClient(request.Request),
rw.Header().Get("Content-Type"), delegate.status, now)
+		Monitor(&verb, &resource,
cleanUserAgent(utilnet.GetHTTPClient(request.Request)),
rw.Header().Get("Content-Type"), delegate.status, now)

Two suggestions.

   1. Could we control this behavior with a flag? I can imagine admins
   might want to turn it off to debug something.
   2. Can we add an extra label to the metric explaining whether or not the
   client field has been canonicalized? I vote for a string clientType
   which is either raw or canonicalized, but we could also do this with a
   bool.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/31783#pullrequestreview-18748752>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p5KdJN7w2vseHEjr1-WkDN-n5idbks5rWSoRgaJpZM4JxlQj>
.

smarterclayton commented:
I am 100% behind getting cardinality down.  I don't think reporting the exact value is useful or desired - this is for metrics and general use, not specific tracking of every user on the platform.

dominikschulz committed:
Clean user agent

This commit cleans common browser user-agents to reduce the metrics
cardinality in exported prometheus metrics.

dominikschulz commented:
Not sure if those failures are real failures or flaky tests. Please advise.

deads2k commented:
the dreaded bazel. `hack/update-bazel.sh`

dominikschulz commented:
Rebased onto master

dominikschulz commented:
@mml @smarterclayton I'd prefer not to introduce an additonal config option/switch to tune this behaviour, instead I'd prefer to keep it as is. WDYT?

smarterclayton commented:
Reducing cardinality is the correct thing to do.  This is not supposed to be a replacement for proxy http logging.  I'm ok with not making this configurable.

dominikschulz commented:
@mml @smarterclayton Rebased. Please advise on how to continue.

smarterclayton commented:
One more time 

/approve

dominikschulz commented:
Rebased.

I'm starting to thing that Bazel isn't really suitable for larger projects like this ...

smarterclayton commented:
/lgtm

smarterclayton commented:
Considering this a bug in that high cardinality causes problems for metrics systems (including those monitoring the cluster), we have unbounded scale there, and this has been on our target list for two releases.  Low risk.

dominikschulz commented:
Thanks!
